include/rbdl/Joint.h

```diff
@@ -481,25 +481,31 @@ struct RBDL_DLLAPI Joint {
     
     // TODO this has to be properly determined AND test case. Try Matt's dot product idea
 #ifdef RBDL_USE_CASADI_MATH
-    if (!axis_0[0].is_zero()) {
+    if (fabs(fabs(casadi::MX::evalf(axis_0[0])) - 1.0).scalar() <
+	std::numeric_limits<double>::epsilon()) {
 #else
     if (axis_0 == Math::SpatialVector(1., 0., 0., 0., 0., 0.)) {
 #endif
       mJointType = JointTypeRevoluteX;
 #ifdef RBDL_USE_CASADI_MATH
-    } else if (!axis_0[1].is_zero()) {
+    } else if (fabs(fabs(casadi::MX::evalf(axis_0[1])) - 1.0).scalar() <
+	       std::numeric_limits<double>::epsilon()){
 #else
     } else if (axis_0 == Math::SpatialVector(0., 1., 0., 0., 0., 0.)) {
 #endif
       mJointType = JointTypeRevoluteY;
 #ifdef RBDL_USE_CASADI_MATH
-    } else if (!axis_0[2].is_zero()) {
+    } else if (fabs(fabs(casadi::MX::evalf(axis_0[2])) - 1.0).scalar() <
+	       std::numeric_limits<double>::epsilon())  {
 #else
     } else if (axis_0 == Math::SpatialVector(0., 0., 1., 0., 0., 0.)) {
 #endif
       mJointType = JointTypeRevoluteZ;
 #ifdef RBDL_USE_CASADI_MATH
-    } else if (axis_0[0].is_zero() && axis_0[1].is_zero() && axis_0[2].is_zero()) {
+    } else if (
+	       fabs(casadi::MX::evalf(axis_0[0]).scalar() < std::numeric_limits<double>::epsilon()) &&
+	       fabs(casadi::MX::evalf(axis_0[1]).scalar() < std::numeric_limits<double>::epsilon()) &&
+	       fabs(casadi::MX::evalf(axis_0[2]).scalar() < std::numeric_limits<double>::epsilon())) {
```

Comment:
I also not sure why this change of condition would be necessary? But I am not a casadi expert, so maybe @pariterre  could you have a look at this?

Reply:
Because is_zero() checks if it is a pure 0 (as opposed to a numerical zero which is not possible with casadi). Therefore, it cannot be evaluated the same. is_zero is therefore the right way to test this part

Reply:
okay so this change set is wrong then …

Reply:
I've never done it that way (evalf)... But since that MX is necessary a 0 or a 1, it may work. I am not sure how it is better that "is_zero()" though

Reply:
I am all for the code being more readable. And is_null seams less confusing to me. Unless there is a specific reason this was changed @PeterQLee?

Reply:
you meant `is_zero`?

Reply:
yes is_zero :laughing:

Reply:
`is_zero` doesn't actually work. I believe that `is_zero` in Casadi is used to check the sparsity of the vector/matrix, and it seems that if you create a vector in a non-sparse way (e.g., `SpatialVector(1., 0., 0., 0., 0., 0.)`) the condition will fail.

Even in the Casadi code they note that "false negatives" are possible. https://github.com/casadi/casadi/blob/master/casadi/core/mx.hpp#L275

The only definitive way I've figured out how to handle this is to a) do all the logic while the values are still floating point and before they've been converted to a symbolic or b) use evalf to find the floating point value.

Reply:
Okay that settle is then :+1: